frameworks/moveos-stdlib/sources/event_queue.move

```diff
@@ -0,0 +1,260 @@
+
+module moveos_std::event_queue {
+    
+    use std::vector;
+    use std::option::{Self, Option};
+    use std::string::String;
+    use moveos_std::timestamp;
+    use moveos_std::object::{Self, Object, ObjectID};
+
+    const MAX_SUBSCRIBER_COUNT: u64 = 100;
+    const MAX_EVENT_HISTORY: u64 = 1000;
+    const SUBSCRIBERS_KEY: vector<u8> = b"subscribers";
+
+    const ErrorTooManySubscribers: u64 = 1;
+    const ErrorSubscriberNotFound: u64 = 2;
+    const ErrorEventNotFound: u64 = 3;
+
+    struct EventQueue<phantom E> has key{
+        sequence_number: u64,
+    }
+
+    struct OnChainEvent<E> has store, drop {
+        event: E,
+        emit_time: u64,
+        subscriber_count: u64,
+        consumed_count: u64,
+    }
+
+    struct Subscriber<phantom E> has key{
+        /// The event queue name
+        queue_name: String,
+        /// The subscriber consume sequence number
+        sequence_number: u64,
+    }    
+
+    fun event_queue<E : copy + drop + store>(name: String): &mut Object<EventQueue<E>> {
+        let object_id = object::custom_object_id<String, EventQueue<E>>(name);
+        if (!object::exists_object(object_id)) {
+            let event_queue_obj = object::new_with_id(name, EventQueue<E> {
+                sequence_number: 0,
+            });
+            let subscribers : vector<ObjectID> = vector::empty();
+            object::add_field(&mut event_queue_obj, SUBSCRIBERS_KEY, subscribers);
+            //We transfer the display object to the moveos_std
+            //And the caller do not need to care about the event queue object
+            object::transfer_extend(event_queue_obj, @moveos_std);
+        };
+        object::borrow_mut_object_extend<EventQueue<E>>(object_id)
+    }
+
+    fun borrow_subscribers<E>(event_queue_obj: &Object<EventQueue<E>>) : &vector<ObjectID>{
+        object::borrow_field(event_queue_obj, SUBSCRIBERS_KEY)
+    }
+
+    fun borrow_mut_subscribers<E>(event_queue_obj: &mut Object<EventQueue<E>>) : &mut vector<ObjectID>{
+        object::borrow_mut_field(event_queue_obj, SUBSCRIBERS_KEY)
+    }
+
+    #[private_generics(E)]
+    public fun emit<E : copy + drop + store>(name: String, event: E) {
+        let event_queue_obj = event_queue<E>(name);
+        let sequence_number = object::borrow(event_queue_obj).sequence_number;
+        let subscribers = borrow_subscribers(event_queue_obj);
+        //If there is no subscriber, we do not need to store the event
+        if (!vector::is_empty(subscribers)) {
+            let now = timestamp::now_milliseconds();
+            let on_chain_event = OnChainEvent {
+                event: event,
+                emit_time: now,
+                subscriber_count: vector::length(subscribers),
+                consumed_count: 0,
+            };
+            object::add_field(event_queue_obj, sequence_number, on_chain_event);
+            //Remove the oldest event if the event history is full
+            if (sequence_number > MAX_EVENT_HISTORY) {
```

Comment:
The event queue has a size limit, and consumption may not be timely, resulting in data missing. For Bitseed, mint may occur and there is no index.

Reply:
What serious problems will occur if the event queue size is not limited?

Reply:
storage costs, we need find a way to remove the old event, if the subscriber doesn't consume it.

Reply:
It seems more appropriate to use an expiration time, so that there will be no problem of event loss due to a large number of writes.

Reply:
We can clean up the expired time every time we write, and provide a public method to clean up expired events.